hashtable.go

```diff
@@ -0,0 +1,186 @@
+		p := pos.partition
+		partitionRecords[p] = append(partitionRecords[p], pos)
+	}
+
+	g, ctx := errgroup.WithContext(context.Background())
+	for i := range partitionRecords {
+		partition := i
+		if len(partitionRecords[partition]) == 0 {
+			continue
+		}
+		g.Go(func() error {
+			// get the bolt db instance for this partition
+			table := ht.tables[partition]
+			for _, record := range partitionRecords[partition] {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				default:
+					if record.key == nil || len(record.key) == 0 {
+						return errors.New("key required")
+					}
+					matchKey := func(slot diskhash.Slot) (bool, error) {
+						if murmur3.Sum32(record.key) == slot.Hash {
+							return true, nil
+						}
+						return false, nil
+					}
+					encPos := encodePosition(record.position)
+					if err := table.Put(record.key, encPos, matchKey); err != nil {
+						return err
+					}
+				}
+			}
+			return nil
+		})
+	}
+	return g.Wait()
+}
+
+// Get chunk position by key
+func (ht *HashTable) Get(key []byte) (*KeyPosition, error) {
+	p := ht.options.getKeyPartition(key)
+	table := ht.tables[p]
+	var value []byte
+	var keyPos *KeyPosition
+
+	getFunc := func(slot diskhash.Slot) (bool, error) {
+		if murmur3.Sum32(key) == slot.Hash {
+			value = make([]byte, len(slot.Value))
+			copy(value, slot.Value)
+			return true, nil
+		}
+		return false, nil
+	}
+
+	err := table.Get(key, getFunc)
+	if err != nil {
+		return nil, err
+	}
+	if len(value) != 0 {
+		keyPos = new(KeyPosition)
+		keyPos.key, keyPos.partition = key, uint32(p)
+		keyPos.position = wal.DecodeChunkPosition(value)
+	}
+	return keyPos, nil
+}
+
+// DeleteBatch delete batch records from index
+func (ht *HashTable) DeleteBatch(keys [][]byte) error {
+	if len(keys) == 0 {
+		return nil
+	}
+	partitionKeys := make([][][]byte, ht.options.partitionNum)
+	for _, key := range keys {
+		p := ht.options.getKeyPartition(key)
+		partitionKeys[p] = append(partitionKeys[p], key)
+	}
+	g, ctx := errgroup.WithContext(context.Background())
+	for i := range partitionKeys {
+		partition := i
+		if len(partitionKeys[partition]) == 0 {
+			continue
+		}
+		g.Go(func() error {
+			table := ht.tables[partition]
+			for _, key := range partitionKeys[partition] {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				default:
+					matchKey := func(slot diskhash.Slot) (bool, error) {
+						if murmur3.Sum32(key) == slot.Hash {
+							return true, nil
+						}
+						return false, nil
+					}
+					if err := table.Delete(key, matchKey); err != nil {
+						return err
+					}
+				}
+			}
+			return nil
+		})
+	}
+	return g.Wait()
+}
+
+// Sync sync index data to disk
+func (ht *HashTable) Sync() error {
+	return nil
+}
+
+// Close index
+func (ht *HashTable) Close() error {
+	for _, table := range ht.tables {
+		err := table.Close()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func encodePosition(cp *wal.ChunkPosition) []byte {
```

Comment:
`ChunkPosition` already has a function named Encode

Reply:
`wal.ChunkPostion.Encode()` is not of fixed size. We can use it but need to make slice twice and copy, like 
```
encPos := make([]byte, binary.MaxVarintLen32*3+binary.MaxVarintLen64)
copy(encPos, record.position.Encode())
```

Reply:
It`s ok, we can use `bytebufferpool` to avoid allocating every time.

Two similar functions seem a little weird.